<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>FAQ • TCRconvertR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="FAQ">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TCRconvertR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/TCRconvertR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/contributing.html">Contributing</a></li>
    <li><a class="dropdown-item" href="../articles/custom-species.html">Using a custom species</a></li>
    <li><a class="dropdown-item" href="../articles/faq.html">FAQ</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/seshadrilab/tcrconvertr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>FAQ</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/seshadrilab/tcrconvertr/blob/main/vignettes/articles/faq.Rmd" class="external-link"><code>vignettes/articles/faq.Rmd</code></a></small>
      <div class="d-none name"><code>faq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="how-does-tcrconvertr-work">How does TCRconvertR work?<a class="anchor" aria-label="anchor" href="#how-does-tcrconvertr-work"></a>
</h2>
<p>TCRconvertR essentially performs a <code>merge</code> between the
input data and a lookup table that includes the naming conventions for
each gene. These lookup tables are constructed from IMGT reference FASTA
files and account for the specific naming peculiarities of each
platform. The built-in lookup tables are located under
<code>inst/extdata/</code>.</p>
<p>TCRconvert performs a <code>merge</code> between the input data and a
lookup table that includes gene names with each nomenclature. These
lookup tables are constructed from IMGT reference FASTA files and
account for the specific naming peculiarities of each platform. The
built-in lookup tables are located under <code>inst/extdata/</code>. The
code used to build the lookup tables, which demonstrates the conversion
logic, is within the <a href="https://github.com/seshadrilab/tcrconvert/blob/361f930d598996f3fa239a715c829ad8c107c63e/tcrconvert/build_lookup.py#L196" class="external-link">build_lookup_from_fastas</a>
function.</p>
</div>
<div class="section level2">
<h2 id="what-input-columns-are-required">What input columns are required?<a class="anchor" aria-label="anchor" href="#what-input-columns-are-required"></a>
</h2>
<p>TCRconvertR expects at least one V, D, J, and/or C gene column. You
can use standard 10X and Adaptive column names or custom names.</p>
</div>
<div class="section level2">
<h2 id="what-if-i-have-missing-genes">What if I have missing genes?<a class="anchor" aria-label="anchor" href="#what-if-i-have-missing-genes"></a>
</h2>
<p><code>NA</code> values in the input dataframe will remain
<code>NA</code> in the output. Genes that are not found in the lookup
table (which is based on the IMGT reference), will be converted to
<code>NA</code>. The built-in lookup tables are located under
<code>inst/extdata/</code>.</p>
</div>
<div class="section level2">
<h2 id="are-gamma-delta-tcrs-supported">Are gamma-delta TCRs supported?<a class="anchor" aria-label="anchor" href="#are-gamma-delta-tcrs-supported"></a>
</h2>
<p>Yes, for human, mouse, and rhesus macaque.</p>
</div>
<div class="section level2">
<h2 id="how-are-alleles-added-from-10x-data">How are alleles added from 10X data?<a class="anchor" aria-label="anchor" href="#how-are-alleles-added-from-10x-data"></a>
</h2>
<p>Since 10X does not provide allele-level information, all genes are
assigned the allele <code>*01</code>.</p>
</div>
<div class="section level2">
<h2 id="how-are-c-genes-converted-to-adaptive">How are C genes converted to Adaptive?<a class="anchor" aria-label="anchor" href="#how-are-c-genes-converted-to-adaptive"></a>
</h2>
<p>Adaptive does not capture constant (“C”) gene information. If
converting to the Adaptive format, all C genes will be set to
<code>NA</code>.</p>
</div>
<div class="section level2">
<h2 id="what-column-names-should-i-use-for-my-imgt-formatted-data">What column names should I use for my IMGT-formatted data?<a class="anchor" aria-label="anchor" href="#what-column-names-should-i-use-for-my-imgt-formatted-data"></a>
</h2>
<p>IMGT does not have standard column names, so it’s assumed that the
10X names are used: c(<code>v_gene</code>, <code>d_gene</code>,
<code>j_gene</code>, <code>c_gene</code>). To use other names, specify
them as a character vector with <code>frm_cols</code>.</p>
</div>
<div class="section level2">
<h2 id="can-i-input-airr-files">Can I input AIRR files?<a class="anchor" aria-label="anchor" href="#can-i-input-airr-files"></a>
</h2>
<p>Yes, just specify the AIRR column names c(<code>v_call</code>,
<code>d_call</code>, <code>j_call</code>, <code>c_call</code>) using
<code>frm_cols</code>. You must still specify the input naming
convention with <code>frm</code>.</p>
</div>
<div class="section level2">
<h2 id="what-if-i-have-custom-column-names">What if I have custom column names?<a class="anchor" aria-label="anchor" href="#what-if-i-have-custom-column-names"></a>
</h2>
<p>If you’re using non-standard column names that do not match 10X,
Adaptive, or Adaptive V2 formats, specify them with
<code>frm_cols</code>.</p>
</div>
<div class="section level2">
<h2 id="what-about-odd-names-e-g--trav14dv4-tcrav01-0212-02">What about odd names (e.g. <code>TRAV14DV4</code>,
<code>TCRAV01-02/12-02</code>)?<a class="anchor" aria-label="anchor" href="#what-about-odd-names-e-g--trav14dv4-tcrav01-0212-02"></a>
</h2>
<p>Gene names containing “OR” or “DV” are accounted for in the lookup
tables.</p>
<p>Combinations of gene names, like <code>TCRAV01-02/12-02</code>, will
be converted to <code>NA</code> because they are not in the IMGT
reference.</p>
</div>
<div class="section level2">
<h2 id="are-non-human-species-supported">Are non-human species supported?<a class="anchor" aria-label="anchor" href="#are-non-human-species-supported"></a>
</h2>
<p>Mouse and rhesus macaque are supported out-of-the-box. For other
species, see “Using a custom reference” in the usage pages.</p>
<p>The rhesus and mouse lookup tables were built from IMGT reference
FASTAs and gene tables. Mouse genes cover both “Mouse” and “Mouse
C57BL/6J” as listed in IMGT.</p>
</div>
<div class="section level2">
<h2 id="what-if-my-adaptive-data-lacks-x_resolvedxmaxresolved-columns">What if my Adaptive data lacks
<code>x_resolved</code>/<code>xMaxResolved</code> columns?<a class="anchor" aria-label="anchor" href="#what-if-my-adaptive-data-lacks-x_resolvedxmaxresolved-columns"></a>
</h2>
<p>Create them by combining <code>x_gene</code>/<code>xGeneName</code>
and <code>x_allele</code>/<code>xGeneAllele</code> with <code>*</code>
as a separator. Example code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adaptive</span></span>
<span><span class="va">new_df</span> <span class="op">&lt;-</span> <span class="va">adaptive_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>v_resolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>    <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">v_allele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">v_gene</span>, <span class="st">"*"</span>, <span class="va">v_allele</span><span class="op">)</span>,</span>
<span>    .default <span class="op">=</span> <span class="va">v_gene</span></span>
<span>    <span class="op">)</span>,</span>
<span>    d_resolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">d_allele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">d_gene</span>, <span class="st">"*"</span>, <span class="va">d_allele</span><span class="op">)</span>,</span>
<span>      .default <span class="op">=</span> <span class="va">d_gene</span></span>
<span>    <span class="op">)</span>,</span>
<span>    j_resolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">j_allele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">j_gene</span>, <span class="st">"*"</span>, <span class="va">j_allele</span><span class="op">)</span>,</span>
<span>      .default <span class="op">=</span> <span class="va">j_gene</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Adaptive v2</span></span>
<span><span class="va">new_df</span> <span class="op">&lt;-</span> <span class="va">adaptive_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>vMaxResolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>    <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">vGeneAllele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">vGeneName</span>, <span class="st">"*"</span>, <span class="va">vGeneAllele</span><span class="op">)</span>,</span>
<span>    .default <span class="op">=</span> <span class="va">vGeneName</span></span>
<span>    <span class="op">)</span>,</span>
<span>    dMaxResolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">dGeneAllele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dGeneName</span>, <span class="st">"*"</span>, <span class="va">dGeneAllele</span><span class="op">)</span>,</span>
<span>      .default <span class="op">=</span> <span class="va">dGeneName</span></span>
<span>    <span class="op">)</span>,</span>
<span>    jMaxResolved <span class="op">=</span> <span class="fu">case_when</span><span class="op">(</span></span>
<span>      <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">jGeneAllele</span><span class="op">)</span> <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">jGeneName</span>, <span class="st">"*"</span>, <span class="va">jGeneAllele</span><span class="op">)</span>,</span>
<span>      .default <span class="op">=</span> <span class="va">jGeneName</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Emma Bishop.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
